plutus-emulator/src/Wallet/Emulator/MultiAgent.hs

```diff
@@ -0,0 +1,220 @@
+
+walletClientState :: Wallet.Wallet -> Lens' EmulatorState NC.NodeClientState
+walletClientState wallet = walletClientStates . at wallet . non NC.emptyNodeClientState
+
+-- | Get the blockchain as a list of blocks, starting with the oldest (genesis)
+--   block.
+chainOldestFirst :: Lens' EmulatorState Blockchain
+chainOldestFirst = chainState . Chain.chainNewestFirst . reversed
+
+-- | Get a map with the total value of each wallet's "own funds".
+fundsDistribution :: EmulatorState -> Map Wallet.Wallet Value
+fundsDistribution = Map.map (foldMap (txOutValue . txOutTxOut) . view Wallet.ownFunds) . view walletStates
+
+-- | Get the emulator log.
+emLog :: EmulatorState -> [EmulatorEvent]
+emLog = view emulatorLog
+
+emptyEmulatorState :: EmulatorState
+emptyEmulatorState = EmulatorState {
+    _chainState = Chain.emptyChainState,
+    _walletStates = mempty,
+    _walletClientStates = mempty,
+    _emulatorLog = mempty
+    }
+
+-- | Initialise the emulator state with a blockchain.
+emulatorState :: Blockchain -> EmulatorState
+emulatorState bc = emptyEmulatorState
+    & chainState . Chain.chainNewestFirst .~ bc
+    & chainState . Chain.index .~ Index.initialise bc
+
+-- | Initialise the emulator state with a pool of pending transactions.
+emulatorStatePool :: Chain.TxPool -> EmulatorState
+emulatorStatePool tp = emptyEmulatorState
+    & chainState . Chain.txPool .~ tp
+
+-- | Initialise the emulator state with a single pending transaction that
+--   creates the initial distribution of funds to public key addresses.
+emulatorStateInitialDist :: Map PubKey Value -> EmulatorState
+emulatorStateInitialDist mp = emulatorStatePool [tx] where
+    tx = Tx
+            { txInputs = mempty
+            , txOutputs = uncurry (flip pubKeyTxOut) <$> Map.toList mp
+            , txForge = foldMap snd $ Map.toList mp
+            , txFee = mempty
+            , txValidRange = WAPI.defaultSlotRange
+            , txSignatures = mempty
+            , txData = mempty
+            }
+
+type MultiAgentEffs = '[State EmulatorState, Error WAPI.WalletAPIError, Error AssertionError, Chain.ChainEffect]
+
+handleMultiAgent
+    :: forall effs. Members MultiAgentEffs effs
+    => Eff (MultiAgentEffect ': effs) ~> Eff effs
+handleMultiAgent = interpret $ \case
+    -- TODO: catch, log, and rethrow wallet errors?
+    WalletAction wallet act -> act
+        & raiseEnd2
+        & Wallet.handleWallet
+        & subsume
+        & NC.handleNodeClient
+        & interpret (handleZoomedState (walletState wallet))
+        & interpret (handleZoomedWriter p1)
+        & interpret (handleZoomedState (walletClientState wallet))
+        & interpret (handleZoomedWriter p2)
+        & interpret (writeIntoState emulatorLog)
+        where
+            p1 :: Prism' [EmulatorEvent] [Wallet.WalletEvent]
+            p1 = below (walletEvent wallet)
+            p2 :: Prism' [EmulatorEvent] [NC.NodeClientEvent]
+            p2 = below (walletClientEvent wallet)
```

Comment:
We might use a class here to get rid of the explicit prisms:

```haskell
class a <: s where
    subtype :: Prism' s a

instance NC.NodeClientEvent <: EmulatorEvent
```

Reply:
Don't typeclasses like this where you don't have a fundep behave quite badly?

Also note that you actually need to pass a parameter to get the prism, so I don't think that would work.